tests/python/pants_test/engine/scheduler_test_base.py

```diff
@@ -97,3 +100,120 @@ def execute_raising_throw(self, scheduler, product, subject):
+    `OrderedDict` of 2-tuples may be used if targets need to be created in a specific order (e.g. if
+    they have dependencies on each other).
+    """
+    raise NotImplementedError('dist_specs must be implemented!')
+
+  @classproperty
+  def run_before_task_types(cls):
+    """
+    By default, we just use a `BuildLocalPythonDistributions` task. When testing with C/C++ targets,
+    we want to compile and link them as well to get the resulting dist to build, so we add those
+    task types here and execute them beforehand.
+    """
+    return []
+
+  @classproperty
+  def run_after_task_types(cls):
+    """Tasks to run after local dists are built, similar to `run_before_task_types`."""
+    return []
+
+  def populate_target_dict(self):
+    self.target_dict = {}
+
+    # Create a target from each specification and insert it into `self.target_dict`.
+    for target_spec, target_kwargs in self.dist_specs.items():
+      unprocessed_kwargs = target_kwargs.copy()
+
+      target_base = Address.parse(target_spec).spec_path
+
+      # Populate the target's owned files from the specification.
+      filemap = unprocessed_kwargs.pop('filemap', {})
+      for rel_path, content in filemap.items():
+        buildroot_path = os.path.join(target_base, rel_path)
+        self.create_file(buildroot_path, content)
+
+      # Ensure any dependencies exist in the target dict (`dist_specs` must then be an
+      # OrderedDict).
+      # The 'key' is used to access the target in `self.target_dict`.
+      key = unprocessed_kwargs.pop('key')
+      dep_targets = []
+      for dep_spec in unprocessed_kwargs.pop('dependencies', []):
+        existing_tgt_key = self.dist_specs[dep_spec]['key']
+        dep_targets.append(self.target_dict[existing_tgt_key])
+
+      # Register the generated target.
+      generated_target = self.make_target(
+        spec=target_spec, dependencies=dep_targets, **unprocessed_kwargs)
+      self.target_dict[key] = generated_target
+
+  @memoized_method
+  def _synthesize_task_types(self, task_types=()):
+    return [
+      self.synthesize_task_subtype(tsk, '__tmp_{}'.format(tsk.__name__))
+      # TODO: make @memoized_method convert lists to tuples for hashing!
+      for tsk in task_types
+    ]
+
+  def _all_specified_targets(self):
+    return list(self.target_dict.values())
+
+  def _scheduling_context(self, **kwargs):
+    scheduler = self.mk_scheduler(rules=self.rules())
+    return self.context(scheduler=scheduler, **kwargs)
+
+  def _retrieve_single_product_at_target_base(self, product_mapping, target):
+    product = product_mapping.get(target)
+    base_dirs = list(product.keys())
+    self.assertEqual(1, len(base_dirs))
+    single_base_dir = base_dirs[0]
+    all_products = product[single_base_dir]
+    self.assertEqual(1, len(all_products))
+    single_product = all_products[0]
+    return single_product
+
+  def _create_task(self, task_type, context):
```

Comment:
`TaskTestBase` should handle this I think.

Reply:
Added a docstring -- this is used specifically to hydrate the `run_before_task_types` and `run_after_task_types` in `self.invoke_tasks()`.